### PR TITLE
Always include autograd context id in rpc/remote requests

### LINKED ISSUE #28819: rpc/remote should always include autograd context id if contains UDF
Currently, we only include context id when there is a valid context and when at least one tensor in the arg list requires grad. However, even if the request does not contain any requires_grad tensor, the response might contain some, whose grad should still belong to the same context.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29543 [WIP] Add end-to-end test for RNN Module
* **#29781 Always include autograd context id in rpc/remote requests**
* #29634 Allow rpc.remote to create RRef on self

Even though the request might not contain any requires_grad tensor,
the return value could. Therefore, we should always include the
autograd context id in the request.

closes #28819

Differential Revision: [D18496709](https://our.internmc.facebook.com/intern/diff/D18496709)

### CODE REVIEW

File: test/dist_autograd_test.py [hunk 1]
```diff
@@ -599,6 +600,42 @@ def test_no_graph_with_tensors_not_require_grad(self):
...
+                    ret_requires_grad
+                ).to_here()
+            else:
+                raise ValueError("Unrecognized ExecMode {}".format(exec_mode))
+
+            ctx = dist_autograd._current_context()
+            send_functions = ctx._send_functions()
+            self.assertEqual(len(send_functions), 0)
+            recv_functions = ctx._recv_functions()
+            self.assertEqual(len(recv_functions), 1)
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: Instead of checking the autograd functions, can we run the backward pass and verify the gradient is accumulated correctly on the remote tensor? We probably might need to return a global tensor from `ret_requires_grad` to validate the grads are accumulated for `ExecMode.RPC_SYNC`

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Thanks for fixing this! Just have one comment inline for the unit test.

File: test/dist_autograd_test.py (hunk 1 above)
Comment (mrshenli) [reply test/dist_autograd_test.py]: Yes, let me add that.

**Review by pritamdamania87 — APPROVED** [review]

File: test/dist_autograd_test.py [hunk 2]
```diff
@@ -599,6 +602,41 @@ def test_no_graph_with_tensors_not_require_grad(self):
...
+                raise ValueError("Unrecognized ExecMode {}".format(exec_mode))
+
+            dist_autograd.backward([ret.sum()])
+
+            rpc.rpc_sync("worker{}".format(dst_rank),
+                         _set_rpc_done, args=(context_id, 1))
+
+            # Wait for the prev rank to be done with rpc.
+            self._check_rpc_done(1)
+            grads = dist_autograd.get_gradients(ctx_ids[1])
```
Comment (pritamdamania87) [inline test/dist_autograd_test.py]: nit: can we also check that the size of `grads` is 1?

Comment (mrshenli) [reply test/dist_autograd_test.py]: yes, let me add.

**Comment by mrshenli** [comment]
The lint failure is on a file that I didn't touch:

```
/home/vsts/work/1/s/aten/src/ATen/cuda/ATenCUDAGeneral.h:3:10: error: 'cuda.h' file not found [clang-diagnostic-error]
```
